modules/alloydb/main.tf

```diff
@@ -407,31 +425,96 @@ resource "google_alloydb_instance" "secondary" {
+  for_each       = local.read_pool
+  annotations    = var.annotations
+  cluster        = google_alloydb_cluster.primary.id
+  database_flags = each.value.flags
+  display_name   = coalesce(each.value.display_name, "${local.prefix}${each.key}")
+  instance_id    = "${local.prefix}${each.key}"
+  instance_type  = "READ_POOL"
+  labels         = var.labels
+
+  dynamic "client_connection_config" {
+    for_each = each.value.require_connectors != null || each.value.ssl_mode != null ? [""] : []
+    content {
+      require_connectors = each.value.require_connectors
+      dynamic "ssl_config" {
+        for_each = each.value.ssl_mode != null ? [""] : []
+        content {
+          ssl_mode = each.value.ssl_mode
+        }
+      }
+    }
+  }
 
-  # waiting to fix this issue https://github.com/hashicorp/terraform-provider-google/issues/14944
-  lifecycle {
-    ignore_changes = [
-      network_config
-    ]
+  machine_config {
+    cpu_count    = each.value.machine_config.cpu_count
+    machine_type = each.value.machine_config.machine_type
+  }
+
+  # network_config block should exist only when (outbound) public IP is enabled to prevent Terraform state drift
+  dynamic "network_config" {
+    for_each = each.value.network_config.enable_public_ip ? [""] : []
+    content {
+      dynamic "authorized_external_networks" {
+        for_each = toset(each.value.network_config.authorized_external_networks)
+        content {
+          cidr_range = authorized_external_networks.value
+        }
+      }
+      enable_public_ip = true
+    }
+  }
+
+  # psc_instance_config block should exist only when there are PSC allowed consumer projects to prevent Terraform state drift
+  dynamic "psc_instance_config" {
+    for_each = length(local.allowed_consumer_projects) > 0 ? [""] : []
+    content {
+      allowed_consumer_projects = local.allowed_consumer_projects
+    }
   }
+
+  read_pool_config {
+    node_count = each.value.node_count
+  }
+
+  dynamic "query_insights_config" {
+    for_each = each.value.query_insights_config != null ? [""] : []
+    content {
+      query_string_length     = each.value.query_insights_config.query_string_length
+      record_application_tags = each.value.query_insights_config.record_application_tags
+      record_client_address   = each.value.query_insights_config.record_client_address
+      query_plans_per_minute  = each.value.query_insights_config.query_plans_per_minute
+    }
+  }
+
+  depends_on = [google_alloydb_instance.primary]
 }
 
 resource "random_password" "passwords" {
-  for_each = toset([
-    for k, v in coalesce(var.users, {}) :
-    k
-    if v.password == null
-  ])
+  for_each = {
+    for k, v in var.users :
+    k => v
+    if v.type == "ALLOYDB_BUILT_IN" && v.password == null
+  }
   length  = 16
   special = true
 }
 
 resource "google_alloydb_user" "users" {
-  for_each       = local.users
+  for_each       = var.users
   cluster        = google_alloydb_cluster.primary.id
```

Comment:
Should this be done in the secondary instances too?

Reply:
My assumption is that the secondary cluster/instance is read-only replica, which asynchronously copies all the data including users from the primary, hence user management in the primary cluster is enough. However, I cannot fully test this assumption. Please, could you advise?

Reply:
Makes sense